README
JS-Redirection-Mobile 

CHANGELOG.txt
#Update 29/10/2012:

Version 1.0.0 released - update to the regex for redirection. regexp from http://detectmobilebrowsers.com/ (WURFL)

#Update 25/07/2011:

Version 0.9.5 released - "keep_path", "keep_query" and "tablet_url" parameters been added. Also 'noredirection_param' is used now instead of 'redirection_paramName'

#Update 12/04/2011:

Version 0.8.6 released - JSLint fixes

#Update 04/04/2011:

Version 0.8.5 released - "ipad_redirection" changed into "tablet_redirection" to extend the feature to multiple tablets. "beforeredirection_callback" changed slightly to let avoid the redirection.

#Update 02/04/2011:

Version 0.8 released - Added support for "ipad_redirection" and "beforeredirection_callback" properties. Ipad excluded from the standard redirection

#Update 11/02/2011:

Version 0.7 released - Added support for Nintendo WII and more HTC phones.

#Update 21/01/2011:

Version 0.6 released - IMPORTANT Change in parameters -- "redirection_paramName" will replace "param"

#Update 06/01/2011:

Version 0.5 released

#Update 20/12/2010:

Fixed a critical issue on IE

#Update 21/11/2010:

All main mobile devices have been considered.













Feel free to fork it!

test/tests.js
/*
 * 
 * Mocks incapsulated in this config literal. There are mocks for document and for navigator.
 * 
 */
window.TEST ={
	
	config : {
		noredirection_param : "noredirection",
	
		mobile_prefix : "m",
		
		beforeredirection_callback : ""
	},
	
	sessionStorage : {	
		item : "",
		
		getItem : function(name) {
			return window.TEST.sessionStorage.name;
		},
		
		setItem : function(name, value) {
			window.TEST.sessionStorage.name = value;
		}	
	},
	
	
	mockDocument: {
		location : {
			protocol : "http:",
			href : "http://domain.com/test",
			host  : "domain.com",
			pathname : "",
			search : ""
		},
		
		cookie : "",
		
		referrer : ""
	},
	
	mockIphoneNavigator : {
		userAgent : "Mozilla/5.0 (iPhone; U; CPU iPhone OS 3_0 like Mac OS X; en-us) AppleWebKit/528.18 (KHTML, like Gecko) Version/4.0 Mobile/7A341 Safari/528.16"
	},
	
	mockIpadNavigator : {
		userAgent : "Mozilla/5.0 (iPad; U; CPU OS 3_2 like Mac OS X; en-us) AppleWebKit/531.21.10 (KHTML, like Gecko) Version/4.0.4 Mobile/7B334b Safari/531.21.10"
	},
	
	mockAndroidNavigator : {
		userAgent : "Mozilla/5.0 (Linux; U; Android 0.5; en-us) AppleWebKit/522+ (KHTML, like Gecko) Safari/41"
	},
	
	mockFFDesktopNavigator : {
		userAgent : "Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.9.2.6) Gecko/20100629 Firefox/3.6 ( .NET CLR 3.5.30729)"
	},
	
	mockHTCdesireNavigator : {
		userAgent : "Mozilla/5.0 (Linux; U; Android 2.2; es-es; HTC Desire HD 1.39.161.1 Build/FRF91) AppleWebKit/533.1 (KHTML, like Gecko) Version/4.0 Mobile Safari/533.1"
	},
	
	mockSamsungTabNavigator : {
		userAgent : "Mozilla/5.0 (Linux; U; Android 2.2; en-us; SCH-I800 Build/FROYO) AppleWebKit/533.1 (KHTML, like Gecko) Version/4.0 Mobile Safari/533.1	"
	},
	
	mockHtcNavigator : {
		userAgent : "HTC-3470/1.2 Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; .NET CLR 1.0.3705; .NET CLR 1.1.4322; Tablet PC 1.7; .NET CLR 2.0.50727; InfoPath.2)"
	}
	
};

module("Redirection happening");
 
test ('iPhoneRedirection()', function() {
	window.TEST.mockDocument.location.host = "domain.com";
	window.TEST.config.noredirection_param = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator,  window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://m.domain.com", "Redirection for iPhone not happening");
})

test ('AndroidRedirection()', function() {
	window.TEST.mockDocument.location.host = "domain.com";
	window.TEST.config.noredirection_param = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockAndroidNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://m.domain.com", "Redirection for Android not happening");
})

test ('HTCRedirection()', function() {
	window.TEST.mockDocument.location.host = "domain.com";
	window.TEST.config.noredirection_param = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockHTCdesireNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://m.domain.com", "Redirection for Android not happening");
})

test ('WWWRedirection()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.noredirection_param = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://m.domain.com", "Redirection not happening");
})

test ('RedirectionTo.mobi()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_url = "www.domain.mobi";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.domain.mobi", "Redirection to www.domain.mobi not happening");
})

test ('RedirectionTo/m()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_url = "www.domain.com/m";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.domain.com/m", "Redirection to http://www.domain.com/m not happening");
})

test ('RedirectionTomobile()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_prefix = "mobile";
	window.TEST.config.mobile_url = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://mobile.domain.com", "Redirection to  mobile.domain.com not happening");
})

test ('RedirectionToDifferentUrl()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.mobile_prefix = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.whatever.com/example", "Redirection to http://www.whatever.com/example not happening");
})

test ('RedirectionToHttpstUrl()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_url = "";
	window.TEST.config.mobile_scheme = "https";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "https://m.domain.com", "Redirection to https://m.domain.com not happening");
})

test ('RedirectionToParametertUrl()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.mockDocument.location.search = "?id=111";
	window.TEST.mockDocument.location.pathname = "/mail/";
	window.TEST.config.mobile_scheme = "http";
	window.TEST.config.mobile_url = "m.domain.com";
	window.TEST.config.keep_path = true;
	window.TEST.config.keep_query = true;
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://m.domain.com/mail/?id=111", "Redirection to https://m.domain.com not happening");
})

test ('RedirectionIpad()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.mockDocument.location.search = "";
	window.TEST.mockDocument.location.pathname = "";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.mobile_scheme = "https";
	window.TEST.config.tablet_redirection = "true";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIpadNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "https://www.whatever.com/example", "Redirection to https://m.domain.com not happening");
})

test ('RedirectionHtc()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.mobile_scheme = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockHtcNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.whatever.com/example", "Redirection to https://m.domain.com not happening");
})

test ('RedirectionSamsungTab()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.mobile_scheme = "https";
	window.TEST.config.tablet_redirection = "true";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockSamsungTabNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "https://www.whatever.com/example", "Redirection to https://m.domain.com not happening");
})

test ('RedirectionIpadBecauseOfCallback()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.mobile_scheme = "https";
	window.TEST.config.tablet_redirection = "true";
	window.TEST.config.beforeredirection_callback = (function(){return true ;});
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIpadNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "https://www.whatever.com/example", "Redirection to https://m.domain.com not happening");
})

test ('RedirectionTabletAndMobileUrl()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.config.mobile_url = "m.whatever.com";
	window.TEST.config.mobile_scheme = "https";
	window.TEST.config.tablet_url = "tablet.whatever.com";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIpadNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "https://tablet.whatever.com", "Redirection to https://tablet.whatever.com not happening");
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockHtcNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "https://m.whatever.com", "Redirection to https://m.domain.com not happening");
})

test ('RedirectionKeepPathAndQuery()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.mockDocument.location.pathname = "/path1";
	window.TEST.mockDocument.location.search = "?a=b";
	window.TEST.config.mobile_url = "m.whatever.com";
	window.TEST.config.mobile_scheme = "";
	window.TEST.config.keep_path = true;
	window.TEST.config.keep_query = true;
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockHtcNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://m.whatever.com/path1?a=b", "Redirection to https://m.domain.com not happening");
})

module("Redirection not happening");

test ('RedirectionFromMobileToStandardVersion()', function() {
	window.TEST.mockDocument.location.host = "domain.com";
	window.TEST.config.noredirection_param = "";
	window.TEST.config.mobile_prefix = "mobile";
	window.TEST.config.mobile_url = "";
	window.TEST.config.mobile_scheme = "";
	window.TEST.mockDocument.cookie = "";
	window.TEST.mockDocument.referrer = "http://mobile.domain.com/test";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.sessionStorage.getItem("noredirection") === "true", "Error in no Redirection happening from mobile");
})

test ('NoRedirectionFromDesktop()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.mockDocument.location.href = "http://www.domain.com/page";
	window.TEST.config.noredirection_param = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockFFDesktopNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.domain.com/page", "Error in no redirection happening");
})

test ('NoRedirectionFromMobileBecauseOfSessionStorage()', function() {
	window.TEST.mockDocument.location.host = "domain.com";
	window.TEST.mockDocument.location.href = "http://domain.com/home";
	window.TEST.config.noredirection_param = "noredirection";
	window.TEST.mockDocument.referrer = "http://www.domain.com/test";
	window.TEST.sessionStorage.setItem("noredirection","true");
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	
	expect(1);
	ok (window.TEST.mockDocument.location.href === "http://domain.com/home", "Error in no Redirection happening from mobile");
})

test ('NoRedirectionFromMobileBecauseOfCookie()', function() {
	window.TEST.mockDocument.location.host = "domain.com";
	window.TEST.mockDocument.location.href = "http://www.domain.com/home";
	window.TEST.config.noredirection_param = "noredirection";
	window.TEST.sessionStorage = undefined;
	window.TEST.mockDocument.cookie = window.TEST.config.noredirection_param;
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.domain.com/home", "Error in no Redirection happening from mobile");
})

test ('NoRedirectionFromMobileBecauseOfParameter()', function() {
	window.TEST.mockDocument.location.host = "domain.com";
	window.TEST.mockDocument.location.href = "http://www.domain.com/home";
	window.TEST.mockDocument.location.search = "?noredirection=true";
	window.TEST.config.noredirection_param = "noredirection";
	window.TEST.sessionStorage = undefined;
	window.TEST.mockDocument.cookie = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.domain.com/home", "Error in no Redirection happening from mobile");
})

test ('NoRedirectionIpad()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.mockDocument.location.href = "http://www.domain.com/home";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.mobile_scheme = "https";
	window.TEST.config.tablet_redirection = "";
	window.TEST.config.noredirection_param = "ss";
	window.TEST.config.tablet_url = "";
	window.TEST.sessionStorage = undefined;
	window.TEST.mockDocument.cookie = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIpadNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href !== "https://www.whatever.com/example", "Redirection happening");
	ok (window.TEST.mockDocument.location.href === "http://www.domain.com/home", "Redirection happening");
})

test ('NoRedirectionIpad2()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.mockDocument.location.href = "http://www.domain.com/home";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.mobile_scheme = "https";
	window.TEST.config.tablet_redirection = "false";
	window.TEST.config.noredirection_param = "ee";
	window.TEST.sessionStorage = undefined;
	window.TEST.mockDocument.cookie = "";
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIpadNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href !== "https://www.whatever.com/example", "Redirection happening");
	ok (window.TEST.mockDocument.location.href === "http://www.domain.com/home", "Redirection happening");
})

test ('NoRedirectionBecauseOfCallback()', function() {
	window.TEST.mockDocument.location.host = "http://www.domain.com/home";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.noredirection_param = "aa";
	window.TEST.sessionStorage = undefined;
	window.TEST.mockDocument.cookie = "";
	window.TEST.config.beforeredirection_callback = (function(){return false;});
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator,  window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.domain.com/home", "Redirection for iPhone happening");
})

test ('NoRedirectionBecauseOfCallback2()', function() {
	window.TEST.mockDocument.location.host = "www.domain.com";
	window.TEST.mockDocument.location.href = "http://www.domain.com/page";
	window.TEST.config.mobile_url = "www.whatever.com/example";
	window.TEST.config.mobile_scheme = "https";
	window.TEST.config.noredirection_param = "bb";
	window.TEST.sessionStorage = undefined;
	window.TEST.mockDocument.cookie = "";
	window.TEST.config.beforeredirection_callback = (function(){var a = "hi";});
	SA.redirection_mobile(window.TEST.mockDocument, window.TEST, window.TEST.mockIphoneNavigator, window.TEST.config);
	ok (window.TEST.mockDocument.location.href === "http://www.domain.com/page", "Redirection happening");
})


grunt.js
/* global module:false */
module.exports = function( grunt ) {

	grunt.loadNpmTasks('grunt-jasmine-runner');
	
	grunt.initConfig({
		pkg: '<json:package.json>',

	    meta: {
	      banner: '/*! <%= pkg.title || pkg.name %> - v<%= pkg.version %> - ' +
	        '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
	        '<%= pkg.homepage ? "* " + pkg.homepage + "\n" : "" %>' +
	        '* Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>;' +
	        ' Licensed <%= _.pluck(pkg.licenses, "type").join(", ") %> */'
	    },

	    watch: {
		    files: '<config:lint.files>',
		    tasks: 'lint test'
		},

	    min: {
	      dist: {
	        src: 'src/<%= pkg.name %>.js',
	        dest: 'dist/<%= pkg.name %>.js'
	      }
	    },

 		jasmine : {
	      src: ['spec/redirection-mobile-testable.js'],
	      specs : 'spec/**/*Spec.js'
	    },

		lint: {
	      files: ['src/<%= pkg.name %>.js', 'src/<%= pkg.name %>-self.js']
	    },

	    jshint: {
	      options: {
	        curly: true,
	        eqeqeq: true,
	        immed: true,
	        latedef: true,
	        newcap: true,
	        noarg: true,
	        sub: true,
	        undef: true,
	        boss: true,
	        eqnull: true,
	        browser: true
	      }
	    }
	});

	// Default task.
	grunt.registerTask('default', 'lint jasmine min');
	grunt.registerTask('travis', 'jasmine');
};

redirection_mobile_self.js
/*!
* JS Redirection Mobile
*
* Copyright (c) 2011 Sebastiano Armeli-Battana (http://sebarmeli.com)
* Dual licensed under the MIT or GPL Version 3 licenses.
* @link http://github.com/sebarmeli/JS-Redirection-Mobile-Site/
*/

/* 
* This version is exactly the same as redirection_mobile.js, but it's an anonymous self-executing
* function, and it's using the default values for 'redirection_paramName' ("mobile_redirect"), 'mobile_prefix'("m"),
* 'mobile_url'("") and 'cookie_hour'(1). In this way you can drop this file on your web server with no configuration
* 
* @author Sebastiano Armeli-Battana
* @version 0.9.5 
* 
*/

/*globals window, redirection_mobile, document, navigator, location */
(function(window, document, navigator) {

		// Helper function for adding time to the current date
	var addTimeToDate = function(msec) {

		// Get the current date
		var exdate = new Date();

		// Add time to the date
		exdate.setTime(exdate.getTime() + msec);

		//Return the new Date
		return exdate;

	};

	// Helper function for getting a value from a parameter in the querystring
	var getQueryValue = function(param) {

		if (!param) {
			return;
		}

		var querystring = document.location.search,
			queryStringArray = querystring && querystring.substring(1).split("&"),
			i = 0,
			length = queryStringArray.length;

		for (; i < length; i++) {
			var token = queryStringArray[i],
				firstPart = token && token.substring(0, token.indexOf("="));
			if (firstPart === param ) {
				return token.substring(token.indexOf("=") + 1, token.length);
			}
		}

	};

	// Retrieve the User Agent of the browser
	var agent = navigator.userAgent.toLowerCase(),

		// Constants
		FALSE = "false",

		// parameter to pass in the URL to avoid the redirection
		redirection_param = "mobile_redirect",

		// prefix appended to the hostname
		mobile_prefix = "m",

		// new url for the mobile site domain 
		mobile_url = "",

		// protocol for the mobile site domain 
		mobile_protocol = document.location.protocol,

		// URL host of incoming request
		host = document.location.host,

		// value for the parameter passed in the URL to avoid the redirection
		queryValue = getQueryValue(redirection_param),

		// Compose the mobile hostname considering "mobile_url" or "mobile_prefix" + hostname
		mobile_host = mobile_url ||
			(mobile_prefix + "." + 
				(!!host.match(/^www\./i) ?
					host.substring(4) : 
						host)),

		// Expiry hours for cookie
		cookie_hours = 1,

		// Check if the UA is a mobile one (iphone, ipod, android, blackberry)
		isUAMobile =!!(agent.match(/(iPhone|iPod|blackberry|android 0.5|htc|lg|midp|mmp|mobile|nokia|opera mobi|opera mini|palm|pocket|psp|sgh|smartphone|symbian|treo mini|Playstation Portable|SonyEricsson|Samsung|MobileExplorer|PalmSource|Benq|Windows Phone|Windows Mobile|IEMobile|Windows CE|Nintendo Wii)/i));

	// Check if the referrer was a mobile page (probably the user clicked "Go to full site") or in the 
	// querystring there is a parameter to avoid the redirection such as "?mobile_redirect=false"
	// (in that case we need to set a variable in the sessionStorage or in the cookie)
	if (document.referrer.indexOf(mobile_host) >= 0 || queryValue === FALSE ) {

		if (window.sessionStorage) {
			window.sessionStorage.setItem(redirection_param, FALSE);
		} else {
			document.cookie = redirection_param + "=" + FALSE + ";expires="+
				addTimeToDate(3600*1000*cookie_hours).toUTCString();
		}
	}

	// Check if the sessionStorage contain the parameter
	var isSessionStorage = (window.sessionStorage) ? 
		(window.sessionStorage.getItem(redirection_param) === FALSE) :
			false,
		
		// Check if the Cookie has been set up
		isCookieSet = document.cookie ? 
			(document.cookie.indexOf(redirection_param) >= 0) :
				false;

	
	// Check that User Agent is mobile, cookie is not set or value in the sessionStorage not present
	if (isUAMobile && !(isCookieSet || isSessionStorage)) {
		
		document.location.href = mobile_protocol + "//" + mobile_host;
	}
}(window, document, navigator));

redirection_mobile.js
/*!
* JS Redirection Mobile
*
* Developed by
* Sebastiano Armeli-Battana (@sebarmeli) - http://www.sebastianoarmelibattana.com
* Dual licensed under the MIT or GPL Version 3 licenses.
*/
/*
	Copyright (c) 2011 Sebastiano Armeli-Battana (http://www.sebastianoarmelibattana.com)

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
* This script will cover a basic scenario of full JS mobile redirection.
* The user will be redirected to the mobile version of the site (home page)
* if it's trying to access the site from a mobile device. This check is
* mainly done checking the User-Agent string. 
* The mobile URL will be obtained appending a prefix (default is "m") to 
* the hostname of the current URL. It's used a naked domain to avoid conflict with
* www.
* 
* In some cases the user needs to be redirected to the Desktop version of the site 
* from a mobile device. This happens when the user clicks on a link such as "Go to full site"
* or when there is a specific parameter in the querystring.
*
* In that case a new key/value in sessionStorage (for modern browsers) 
* will be set and until the user doesn't close browser window or tab it will access
* to the desktop version from a mobile device. There is a fallback for old browsers that
* don't support sessionStorage, and it will use a cookie. The cookie will expiry in one hour 
* (default value) or you configure the expiry time.
* 
* To use this function, you need to call it as SA.redirection_mobile(config);
* E.g. SA.redirection_mobile ({noredirection_param : "noredirection", mobile_prefix : "mobile", cookie_hours : "2" })
* or
* E.g. SA.redirection_mobile ({mobile_url : "mobile.whatever.com/example", mobile_sheme : "https" })
* or
* E.g. SA.redirection_mobile ({mobile_prefix : "mobile"})
* or
* E.g. SA.redirection_mobile ({mobile_prefix : "mobile", mobile_scheme : "https", redirection_paramName : "modile_redirect"})
* or
* E.g. SA.redirection_mobile ({mobile_url : "mobile.whatever.com/example", tablet_redirection : "true"})
* or
* E.g. SA.redirection_mobile ({{mobile_url : "mobile.whatever.com/example", beforeredirection_callback : (function(){alert('2')}) }})
* or
* E.g. SA.redirection_mobile ({{mobile_url : "mobile.whatever.com", tablet_url : "tablet.whatever.com")
*
*
* @link http://github.com/sebarmeli/JS-Redirection-Mobile-Site/
* @author Sebastiano Armeli-Battana
* @version 0.9.5
* @date 25/07/2011 
* 
*/
	
/*globals window,document, navigator, SA */
if (!window.SA) {window.SA = {};}

/*
* @param configuration object containing three properties:
*			- mobile_prefix : prefix appended to the hostname (such as "m" to redirect to "m.domain.com")
*			- mobile_url : mobile url to use for the redirection (such as "mobile.whatever.com" to redirect to "whatever.com" )
*			- mobile_scheme : url scheme (http/https) of the mobile site domain
*			- cookie_hours : number of hours the cookie needs to exist after redirection to desktop site
*			- noredirection_param : parameter to pass in the querystring of the URL to avoid the redirection (the value must be equal to "true").
*				It's also the name of the item in the localStorage (or cookie name) to avoid mobile
*				redirection. Default value is "noredirection". Eg: http://domain.com?noredirection=true
*			- tablet_redirection : boolean value that enables/disables(default) the redirection for tablet such as iPad, Samsung Galaxy Tab, Kindle or Motorola Xoom. - Default:false
*				Default Url for redirection will be the same as the one for mobile devices.
*			- tablet_url : url to use for the redirection in case user is using a tablet to access the site
*			- keep_path : boolean to determine if the destination url needs to keep the path from the original url
*			- keep_query : boolean to determine if the destination url needs to keep the querystring from the original url
*			- beforeredirection_callback : callback launched before the redirection happens
*/
SA.redirection_mobile = function(configuration) {

	// Helper function for adding time to the current date
	var addTimeToDate = function(msec) {

		// Get the current date
		var exdate = new Date();

		// Add time to the date
		exdate.setTime(exdate.getTime() + msec);

		//Return the new Date
		return exdate;

	};

	// Helper function for getting a value from a parameter in the querystring
	var getQueryValue = function(param) {

		if (!param) {
			return;
		}

		var querystring = document.location.search,
			queryStringArray = querystring && querystring.substring(1).split("&"),
			i = 0,
			length = queryStringArray.length;

		for (; i < length; i++) {
			var token = queryStringArray[i],
				firstPart = token && token.substring(0, token.indexOf("="));
			if (firstPart === param ) {
				return token.substring(token.indexOf("=") + 1, token.length);
			}
		}

	};
				
	// Retrieve the User Agent of the browser
	var agent = navigator.userAgent.toLowerCase(),
	
		// Constants
		FALSE = "false",
		TRUE = "true",

		// configuration object
		config = configuration || {},
	
		// parameter to pass in the URL to avoid the redirection
		redirection_param = config.noredirection_param || "noredirection",
		
		// prefix appended to the hostname
		mobile_prefix = config.mobile_prefix || "m",
		
		// new url for the mobile site domain 
		mobile_url = config.mobile_url,
		
		// protocol for the mobile site domain 
		mobile_protocol = config.mobile_scheme ?
			config.mobile_scheme + ":" :
				document.location.protocol,
		
		// URL host of incoming request
		host = document.location.host,

		// value for the parameter passed in the URL to avoid the redirection
		queryValue = getQueryValue(redirection_param),

		// Compose the mobile hostname considering "mobile_url" or "mobile_prefix" + hostname
		mobile_host = mobile_url ||
			(mobile_prefix + "." + 
				(!!host.match(/^www\./i) ?
					host.substring(4) : 
						host)),

		// Expiry hours for cookie
		cookie_hours = config.cookie_hours || 1,
		
		// Parameters to determine if the pathname and the querystring need to be kept
		keep_path = config.keep_path || false,
		keep_query = config.keep_query || false,

		// new url for the tablet site domain 
		tablet_host = config.tablet_url || mobile_host,
		
		// Check if the UA is a mobile one (iphone, ipod, android, blackberry)
		isUAMobile =!!(agent.match(/(iPhone|iPod|blackberry|android 0.5|htc|lg|midp|mmp|mobile|nokia|opera mobi|opera mini|palm|pocket|psp|sgh|smartphone|symbian|treo mini|Playstation Portable|SonyEricsson|Samsung|MobileExplorer|PalmSource|Benq|Windows Phone|Windows Mobile|IEMobile|Windows CE|Nintendo Wii)/i));


	// Check if the referrer was a mobile page (probably the user clicked "Go to full site") or in the 
	// querystring there is a parameter to avoid the redirection such as "?mobile_redirect=false"
	// (in that case we need to set a variable in the sessionStorage or in the cookie)
	if (document.referrer.indexOf(mobile_host) >= 0 || queryValue === TRUE ) {

		if (window.sessionStorage) {
			window.sessionStorage.setItem(redirection_param, TRUE);
		} else {
			document.cookie = redirection_param + "=" + TRUE + ";expires="+
				addTimeToDate(3600*1000*cookie_hours).toUTCString();
		}
	}

	// Check if the sessionStorage contain the parameter
	var isSessionStorage = (window.sessionStorage) ? 
			(window.sessionStorage.getItem(redirection_param) === TRUE) :
				false,

		// Check if the Cookie has been set up
		isCookieSet = document.cookie ? 
			(document.cookie.indexOf(redirection_param) >= 0) :
				false;
	
	// Check if the device is a Tablet such as iPad, Samsung Tab, Motorola Xoom or Amazon Kindle
	if (!!(agent.match(/(iPad|SCH-I800|xoom|kindle)/i))) {

		// Check if the redirection needs to happen for iPad
		var isUATablet = (config.tablet_redirection === TRUE || !!config.tablet_url) ? true : false;
		isUAMobile = false;
	}

	// Check that User Agent is mobile, cookie is not set or value in the sessionStorage not present
	if ((isUATablet || isUAMobile) && !(isCookieSet || isSessionStorage)) {

		// Callback call
		if (config.beforeredirection_callback) {
			if (!config.beforeredirection_callback.call(this)) {
				return;
			}
		}
		
		var path_query = "";
		
		if(keep_path) { 
			path_query += document.location.pathname;
		}
		
		if (keep_query) {
			path_query += document.location.search;
		}
		
		if (isUATablet){
			document.location.href = mobile_protocol + "//" + tablet_host + path_query;
		} else if (isUAMobile) {
			document.location.href = mobile_protocol + "//" + mobile_host + path_query;
		}
		
	} 
};	